### PR TITLE
Add Langchain Dependencies to Databricks CreateModelVersion Call 

### PR DESCRIPTION
**Pull request by victorsun123**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/victorsun123/mlflow/pull/11086?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/11086/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 11086
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?
We parsing of Langchain Dependencies from Model Info and adding them as arguments to call to CreateModelVersion in the Databricks UC ModelRegistry.

### How is this PR tested?

- [ ] Existing unit/integration tests
- [X] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [X] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [X] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [X] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [X] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/store/_unity_catalog/registry/rest_store.py [hunk 1]
```diff
@@ -648,6 +660,7 @@ def create_model_version(
                     tags=uc_model_version_tag_from_mlflow_tags(tags),
                     run_tracking_server_id=source_workspace_id,
                     feature_deps=feature_deps,
+                    vector_index_deps = vector_index_deps
```
Comment (kriscon-db) [inline mlflow/store/_unity_catalog/registry/rest_store.py:663]: You'll need to add this to the proto definition in `mlflow/protos/databricks_uc_registry_messages.proto` and rerun `dev/generate-protos.sh`.

File: mlflow/store/_unity_catalog/registry/rest_store.py [hunk 2]
```diff
@@ -160,6 +163,14 @@ def get_feature_dependencies(model_dir):
...
     return ""
 
+def get_vector_search_dependencies(model_dir):
+    model = _load_model(model_dir)
+    model_info = model.get_model_info()
+    index_name = model_info.flavors.get("langchain", {}).get(_DATABRICKS_DEPENDENCY_KEY, {}).get(_DATABRICKS_VECTOR_SEARCH_INDEX_NAME_KEY, None)
+    if index_name:
+        return {"index_name": index_name}
+    else:
+        return None
```
Comment (dbczumar) [inline mlflow/store/_unity_catalog/registry/rest_store.py:173]: Can this information be read directly from the MLmodel file during model version creation? Do we strictly need to extend the CreateModelVersion REST API?

Comment (victorsun123) [reply mlflow/store/_unity_catalog/registry/rest_store.py:173]: Do you know if there's existing precedent for reading from the model file in model registry? How can we access the MlModel file in model registry service?

Comment (dbczumar) [reply mlflow/store/_unity_catalog/registry/rest_store.py:173]: Yes, the model registry service has access to the MLmodel file. For example, it currently uses it to render model signature information in the UI.

@smurching can comment more for UCMR

**Review by aakrati — COMMENTED** [review]
Mostly LGTM to me! the proto field may change based on the PR https://github.com/databricks/universe/pull/489771

File: mlflow/store/_unity_catalog/registry/rest_store.py [hunk 3]
```diff
@@ -648,6 +668,7 @@ def create_model_version(
                     tags=uc_model_version_tag_from_mlflow_tags(tags),
                     run_tracking_server_id=source_workspace_id,
                     feature_deps=feature_deps,
+                    model_version_dependencies = other_model_deps
```
Comment (dbczumar) [inline mlflow/store/_unity_catalog/registry/rest_store.py:671]: @victorsun123 Can we write some unit tests to make sure that we're populating this field correctly?

File: mlflow/store/_unity_catalog/registry/rest_store.py [hunk 4]
```diff
@@ -160,7 +163,23 @@ def get_feature_dependencies(model_dir):
...
     return ""
 
-
+def get_model_version_dependencies(model_dir):
+    model = _load_model(model_dir)
+    model_info = model.get_model_info()
+    dependencies = List()
+    index_name = _fetch_langchain_dependency_from_model_info(model_info, _DATABRICKS_VECTOR_SEARCH_INDEX_NAME_KEY)
+    if index_name:
+        dependencies.append({"kind": "vector_index", "source": index_name})
```
Comment (aakrati) [inline mlflow/store/_unity_catalog/registry/rest_store.py:172]: Lets rename to "name" per offline conversation with @dbczumar

File: mlflow/store/_unity_catalog/registry/rest_store.py (hunk 3 above)
Comment (aakrati) [reply mlflow/store/_unity_catalog/registry/rest_store.py:671]: +1

**Comment by victorsun123** [comment]
@dbczumar got my setup working and tests and lint should be passing! Let me know if there's anything else that needs fixing thanks!

File: mlflow/store/_unity_catalog/registry/rest_store.py [hunk 5]
```diff
@@ -648,6 +679,7 @@ def create_model_version(
                     tags=uc_model_version_tag_from_mlflow_tags(tags),
                     run_tracking_server_id=source_workspace_id,
                     feature_deps=feature_deps,
+                    model_version_dependencies = other_model_deps
```
Comment (dbczumar) [inline mlflow/store/_unity_catalog/registry/rest_store.py:682]: Nit: 
```suggestion
                    model_version_dependencies=other_model_deps
```

Surprised our linter didn't complain

File: mlflow/protos/databricks_uc_registry_messages.proto [hunk 6]
```diff
@@ -274,6 +285,10 @@ message CreateModelVersionRequest {
 
   // Tags associated with model version.
   repeated ModelVersionTag tags = 7;
+
+  // Additional dependencies to pass to createModelVersion including VectorSearch and Embedding endpoints
+  //TODO(ML-38658): migrate feature_deps to under this field as well
```
Comment (dbczumar) [inline mlflow/protos/databricks_uc_registry_messages.proto:290]: Can we remove Databricks ticket numbers from our open source code base?

```suggestion
  // TODO: migrate feature_deps to under this field as well
```

**Review by dbczumar — APPROVED** [review]
LGTM from an MLflow OSS perspective. We'll need @smurching to stamp for UC

File: mlflow/protos/databricks_uc_registry_messages.proto [hunk 7]
```diff
@@ -250,6 +250,17 @@ message SearchRegisteredModelsResponse {
   optional string next_page_token = 2;
 }
 
+enum DependencyType {
+  DEPENDENCY_TYPE_UNSPECIFIED = 0;
+  VECTOR_INDEX = 1;
+  MODEL_ENDPOINT = 2;
```
Comment (smurching) [inline mlflow/protos/databricks_uc_registry_messages.proto:256]:
```suggestion
  DATABRICKS_MODEL_ENDPOINT = 2;
```

Same comment here

**Review by smurching — COMMENTED** [review]
Minor naming comment, otherwise LGTM!

File: mlflow/protos/databricks_uc_registry_messages.proto [hunk 8]
```diff
@@ -250,6 +250,17 @@ message SearchRegisteredModelsResponse {
   optional string next_page_token = 2;
 }
 
+enum DependencyType {
+  DEPENDENCY_TYPE_UNSPECIFIED = 0;
+  VECTOR_INDEX = 1;
```
[the remark this answers is not in the document]
Comment (victorsun123) [reply mlflow/protos/databricks_uc_registry_messages.proto:255]: hmm yeah that's a good point we can make that change but we'll also need to change the proto in UCMR in that case

[the remark this answers is not in the document]
Comment (kriscon-db) [reply mlflow/protos/databricks_uc_registry_messages.proto]: While we should change the name of the field in the UCMR proto to match, this should still work as is.  I believe it uses the index values to map these fields and not the names themselves as defined on the proto.

**Review by smurching — APPROVED** [review]

**Pull request merged**